siliconcompiler/remote/client.py

```diff
@@ -196,30 +283,71 @@ def remote_run(chip):
-                                     int(time.monotonic() - step_start), cur_step))
-                    if cur_log:
-                        chip.logger.info(f"Tail of current logfile:\n{cur_log}\n")
-                else:
-                    chip.logger.info("Job is still running (%d seconds, step: unknown)" % (
-                                     int(time.monotonic() - step_start)))
-        except Exception:
-            # Sometimes an exception is raised if the request library cannot
-            # reach the server due to a transient network issue.
-            # Retrying ensures that jobs don't break off when the connection drops.
-            is_busy = True
-            chip.logger.info("Unknown network error encountered: retrying.")
-    chip.logger.info("Remote job run completed! Fetching results...")
+        all_completed, is_busy = check_progress(chip,
+                                                step_start,
+                                                all_nodes,
+                                                completed)
+        nodes_to_fetch = []
+        for node in all_completed:
+            if node not in completed:
+                nodes_to_fetch.append(node)
+        if nodes_to_fetch:
+            chip.logger.info('  Fetching completed results:')
+            for node in nodes_to_fetch:
+                node_proc = multiprocessor.Process(target=fetch_results,
+                                                   args=(chip, node))
+                node_proc.start()
+                result_procs.append(node_proc)
+                chip.logger.info(f'    {node}')
+        completed = all_completed.copy()
+
+    # Done: try to fetch any node results which still haven't been retrieved.
+    chip.logger.info('Remote job completed! Retrieving final results...')
+    for node in all_nodes:
+        if node not in completed:
+            node_proc = multiprocessor.Process(target=fetch_results,
+                                               args=(chip, node))
+            node_proc.start()
+            result_procs.append(node_proc)
+    # Make sure all results are fetched before letting the client issue
+    # a deletion request.
+    for proc in result_procs:
+        proc.join()
+
+    # Produce final manifest, based on the returned per-node results.
+    # (Un-set the 'remote' option to avoid steplist-based summary/show errors)
+    chip.unset('option', 'remote')
+    chip._finalize_run(chip.list_steps(),
+                       copy.deepcopy(os.environ))
+
+
+###################################
+def check_progress(chip, step_start, all_nodes, completed=[]):
+    try:
+        is_busy_info = is_job_busy(chip)
+        is_busy = is_busy_info['busy']
+        if is_busy:
+            all_completed = _process_progress_info(chip,
+                                                   is_busy_info,
+                                                   step_start,
+                                                   completed)
```

Comment:
the return value is not needed since you can just operate directly on the input argument., that looks to avoid all the copies in this code.